File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int good=0;

    int goodNodes(TreeNode* root) {
        dfs(root, root->val);
        return good;
    }
    void dfs(TreeNode* node, int pathMax){
        if(!node) return;
        if(node->val>=pathMax){
            pathMax=node->val;
            good++;
        }
        dfs(node->left, pathMax);
        dfs(node->right, pathMax);
    }

};

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        inorder_index = {}
        for i, value in enumerate(inorder):
            inorder_index[value] = i

        self.preorder_pos = 0
        
        def dfs(left, right):
            if left > right:
                return None
            root_value = preorder[self.preorder_pos]
            self.preorder_pos += 1

            root = TreeNode(root_value)
            mid = inorder_index[root_value]

            root.left = dfs(left, mid - 1)
            root.right = dfs(mid + 1, right)

            return root

        return dfs(0, len(inorder) - 1)

